Optimize integer `pow` by removing the exit branch

mzabaluev opened the pull request:
The branch at the end of the `pow` implementations is redundant with multiplication code already present in the loop. By rotating the exit check, this branch can be largely removed, improving code size and reducing instruction cache misses.

Testing on my machine (`x86_64`, 11th Gen Intel Core i5-1135G7 @ 2.40GHz), the `num::int_pow` benchmarks improve by some 40% for the unchecked operations and show some slight improvement for the checked operations as well.

mzabaluev committed:
Optimize integer pow by removing exit branch

The branch at the end of the `pow` implementations is redundant
with multiplication code already present in the loop. By rotating
the exit check, this branch can be largely removed, improving code size
and instruction cache coherence.

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 76d253060eca776bc104233253ef54bea8906829 has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu 76d253060eca776bc104233253ef54bea8906829 -->
<!-- homu: {"type":"Approved","sha":"76d253060eca776bc104233253ef54bea8906829","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 76d253060eca776bc104233253ef54bea8906829 with merge 87b8256db8a0bfa95b412f9b5772749ac6d18afe...
<!-- homu: {"type":"BuildStarted","head_sha":"76d253060eca776bc104233253ef54bea8906829","merge_sha":"87b8256db8a0bfa95b412f9b5772749ac6d18afe"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8657859165/job/23740697634) [(plain)](https://github.com/rust-lang-ci/rust/commit/87b8256db8a0bfa95b412f9b5772749ac6d18afe/checks/23740697634/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen/issues/issue-34947-pow-i32.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-34947-pow-i32/issue-34947-pow-i32.ll" "/checkout/tests/codegen/issues/issue-34947-pow-i32.rs" "--check-prefix=CHECK" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
--- stderr -------------------------------
/checkout/tests/codegen/issues/issue-34947-pow-i32.rs:9:17: error: CHECK-NEXT: is not on the line after the previous match
 // CHECK-NEXT: mul
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-34947-pow-i32/issue-34947-pow-i32.ll:27:13: note: 'next' match was here
 %_14.0.i = mul i32 %base.0.i, %base.0.i
            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-34947-pow-i32/issue-34947-pow-i32.ll:20:16: note: previous match ended here
 %_10.0.i = mul i32 %base.0.i, %acc.0.i
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-34947-pow-i32/issue-34947-pow-i32.ll:21:1: note: non-matching line after previous match is here
 %0 = icmp eq i32 %exp.0.i, 1


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-34947-pow-i32/issue-34947-pow-i32.ll


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
        1: ; ModuleID = 'issue_34947_pow_i32.63e30470ddee3881-cgu.0' 
        2: source_filename = "issue_34947_pow_i32.63e30470ddee3881-cgu.0" 
        3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
        4: target triple = "x86_64-unknown-linux-gnu" 
        5:  
        6: ; Function Attrs: nofree norecurse nosync nounwind nonlazybind memory(none) uwtable 
        7: define noundef i32 @issue_34947(i32 noundef %x) unnamed_addr #0 { 
        8: start: 
        9:  br label %bb3.i 
       10:  
       11: bb3.i: ; preds = %bb9.i, %start 
       12:  %acc.0.i = phi i32 [ 1, %start ], [ %acc.1.i, %bb9.i ] 
       13:  %base.0.i = phi i32 [ %x, %start ], [ %_14.0.i, %bb9.i ] 
       14:  %exp.0.i = phi i32 [ 5, %start ], [ %1, %bb9.i ] 
       15:  %_6.i = and i32 %exp.0.i, 1 
       16:  %.not.i = icmp eq i32 %_6.i, 0 
       17:  br i1 %.not.i, label %bb9.i, label %bb4.i 
       18:  
       19: bb4.i: ; preds = %bb3.i 
       20:  %_10.0.i = mul i32 %base.0.i, %acc.0.i 
       21:  %0 = icmp eq i32 %exp.0.i, 1 
       22:  br i1 %0, label %"_ZN4core3num21_$LT$impl$u20$i32$GT$3pow17hdf744fa1bafd0b6cE.exit", label %bb9.i 
       23:  
       24: bb9.i: ; preds = %bb4.i, %bb3.i 
       25:  %acc.1.i = phi i32 [ %_10.0.i, %bb4.i ], [ %acc.0.i, %bb3.i ] 
       26:  %1 = lshr i32 %exp.0.i, 1 
       27:  %_14.0.i = mul i32 %base.0.i, %base.0.i 
next:9                 !~~                           error: match on wrong line
       28:  br label %bb3.i 
       29:  
       30: "_ZN4core3num21_$LT$impl$u20$i32$GT$3pow17hdf744fa1bafd0b6cE.exit": ; preds = %bb4.i 
       31:  ret i32 %_10.0.i 
       32: } 
       33:  
       34: attributes #0 = { nofree norecurse nosync nounwind nonlazybind memory(none) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
       35:  
       36: !llvm.module.flags = !{!0, !1} 
       37: !llvm.ident = !{!2} 
       38:  
       39: !0 = !{i32 8, !"PIC Level", i32 2} 
       40: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
       41: !2 = !{!"rustc version 1.79.0-nightly (87b8256db 2024-04-12)"} 
------------------------------------------



```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8657859165/job/23741164148)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/8657859165/job/23741164148","builder_name":"checks-actions"} -->

mzabaluev commented:
> The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8657859165/job/23740697634) [(plain)](https://github.com/rust-lang-ci/rust/commit/87b8256db8a0bfa95b412f9b5772749ac6d18afe/checks/23740697634/logs)
> Click to see the possible cause of the failure (guessed by this bot)
> 
> ```
> failures:
> 
> ---- [codegen] tests/codegen/issues/issue-34947-pow-i32.rs stdout ----
> 
> error: verification with 'FileCheck' failed
> status: exit status: 1
> command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-34947-pow-i32/issue-34947-pow-i32.ll" "/checkout/tests/codegen/issues/issue-34947-pow-i32.rs" "--check-prefix=CHECK" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
> --- stderr -------------------------------
> --- stderr -------------------------------
> /checkout/tests/codegen/issues/issue-34947-pow-i32.rs:9:17: error: CHECK-NEXT: is not on the line after the previous match
>  // CHECK-NEXT: mul
> ```

I'm not familiar with this check, so I don't understand what's failing here and what should the fix be.

Amanieu commented:
It seems that your PR has introduced a regression: LLVM is no longer able to optimize `pow(5)` down to just 3 multiply instructions.

mzabaluev commented:
> It seems that your PR has introduced a regression: LLVM is no longer able to optimize `pow(5)` down to just 3 multiply instructions.

Does this mean the modified code performs worse in this specific case?

Amanieu commented:
Yes, it will perform much worse in that specific case since LLVM is unable to optimize the loop away. See https://godbolt.org/z/nMY79Gn8r for the current code that is being generated. It might be possible to re-arrange the code so that you still get the performance benefit of this PR while still letting LLVM optimize the loop, but I'm not sure.

Amanieu commented:
You can see a comparison of the old and new versions here: https://godbolt.org/z/dx3WKxhad

RalfJung commented:
@bors r-
(bors sync fixup)

oskgo commented:
@mzabaluev Any updates on this? Thanks!

mzabaluev commented:
The lack of optimization in case of a small const argument value is unfortunate.

I briefly tried to salvage it by giving the optimizer an easier time without re-introducing redundancy in the dynamic case, but didn't come up with any good ideas.

Maybe an unrolled fast path for argument values in the 0..=6 range? This would feel like an exercise in tricking the optimizer and placating the benchmarks.

oskgo commented:
If I understand correctly you don't know how to fix the regression, and you're not going to make the argument that the regression is tolerable?

If I'm right you should probably close this. You can always reopen if you get some new inspiration or can find guidance.

Amanieu commented:
If might be worth trying something with [`is_val_statically_known`](https://doc.rust-lang.org/nightly/std/intrinsics/fn.is_val_statically_known.html) to have 2 different paths depending on whether the input argument is a constant.

mzabaluev committed:
Explicitly unroll integer pow for small exponents

The newly optimized loop has introduced a regression in the case
when pow is called with a small constant exponent. LLVM is no longer
able to unroll the loop and the generated code is larger and slower
than what's expected in tests.

Match and handle small exponent values separately by branching out
to an explicit multiplication sequence for that exponent.
Powers larger than 6 need more than three multiplications, so these
cases are less likely to benefit from this optimization, also such
constant exponents are less likely to be used in practice.
For uses with a non-constant exponent, this might also provide
a performance benefit if the exponent is small and does not vary
between successive calls, so the same match arm tends to be taken as
a predicted branch.

mzabaluev commented:
> If might be worth trying something with [`is_val_statically_known`](https://doc.rust-lang.org/nightly/std/intrinsics/fn.is_val_statically_known.html) to have 2 different paths depending on whether the input argument is a constant.

I will combine this with my suggestion for the statically known case, thanks for the tip!

@oskgo it looks like we've found a way to resolve the regression, don't close this yet.

mzabaluev commented:
I get this error when trying to use `is_val_statically_known` inside `pow` methods:

```
error: `is_val_statically_known` is not yet stable as a const fn
```

mzabaluev commented:
> an unrolled fast path for argument values in the 0..=6 range

This idea seems to work out: the benchmarks are mostly better than on the previous commit, even in variable exponent benchmarks, where the compiler generates jump tables.

oskgo commented:
You can use the `is_val_statically_known` feature to get access to `is_val_statically_known` in const fns. There's a hint in the error message showing you the syntax.

mzabaluev commented:
> You can use the `is_val_statically_known` feature to get access to `is_val_statically_known` in const fns. There's a hint in the error message showing you the syntax.

I got no hint (with the bootstrap beta compiler), but regardless, there seems to be no need for this accordingly to the benchmarks: the revision with unrolled multiplications for small exponents performs better in the dynamic case as well. I can also check on an M3 Mac.

oskgo commented:
Right. It's probably because of the bootstrap compiler. I get the [hint on nightly](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=64f83053278360264bd8c3a4f4b35ef7).

I just wanted to make clear that it's possible to use `is_val_statically_known` in const functions. Whether this should be done or not is not up to me. I'm just here to get the PR moving again.

Amanieu commented:
I would still be interested in seeing a version based on `is_val_statically_known`. I'm concerned about the code size impact of the unrolled loop.

mzabaluev commented:
> I would still be interested in seeing a version based on `is_val_statically_known`. I'm concerned about the code size impact of the unrolled loop.

I was thinking of making two conditionally compiled variants, one using `is_val_statically_known` for targets that don't do well on the unrolled match in the dynamic exponent case, and the other without it, for `x86_64` and possibly others.
But if you prefer to trade some performance for reduced code size, I'll stick to the single implementation.

mzabaluev committed:
Use is_val_statically_known to optimize pow

In the dynamic exponent case, it's preferred to not increase code size,
so use solely the loop-based implementation there.
This shows about 4% penalty in the variable exponent benchmarks
on x86_64.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9913383995/job/27390283799) [(plain)](https://github.com/rust-lang/rust/commit/010c332555dc0fd79d02c7b04a5cfac40f0af279/checks/27390283799/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#12 [ 5/11] RUN npm install es-check@6.1.1 eslint@8.6.0 -g
#12 5.103 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 5.165 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 5.167 npm WARN deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
#12 5.217 npm WARN deprecated @humanwhocodes/config-array@0.9.5: Use @eslint/config-array instead
#12 5.239 npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
#12 5.247 npm WARN deprecated @humanwhocodes/object-schema@1.2.1: Use @eslint/object-schema instead
#12 6.270 
#12 6.270 added 205 packages in 6s
#12 6.271 
#12 6.271 20 packages are looking for funding
---
#16 3.745 Building wheels for collected packages: reuse
#16 3.746   Building wheel for reuse (pyproject.toml): started
#16 4.086   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 4.087   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#16 4.087   Stored in directory: /tmp/pip-ephem-wheel-cache-2s2ro5rw/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#16 4.090 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#16 4.113   Attempting uninstall: setuptools
#16 4.114     Found existing installation: setuptools 59.6.0
#16 4.115     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
    --> library/core/src/num/int_macros.rs:2749:16
     |
1    | /  macro_rules! int_impl {
2    | |      (
3    | |          Self = $SelfT:ty,
4    | |          ActualT = $ActualT:ident,
...    |
2749 | |              if intrinsics::is_val_statically_known(exp) {
...    |
3646 | |      }
3647 | |  }
     | |__- in this expansion of `int_impl!`
---
    --> library/core/src/num/int_macros.rs:2749:16
     |
1    | /  macro_rules! int_impl {
2    | |      (
3    | |          Self = $SelfT:ty,
4    | |          ActualT = $ActualT:ident,
...    |
2749 | |              if intrinsics::is_val_statically_known(exp) {
...    |
3646 | |      }
3647 | |  }
     | |__- in this expansion of `int_impl!`
---
    --> library/core/src/num/int_macros.rs:2749:16
     |
1    | /  macro_rules! int_impl {
2    | |      (
3    | |          Self = $SelfT:ty,
4    | |          ActualT = $ActualT:ident,
...    |
2749 | |              if intrinsics::is_val_statically_known(exp) {
...    |
3646 | |      }
3647 | |  }
     | |__- in this expansion of `int_impl!`
---
    --> library/core/src/num/int_macros.rs:2749:16
     |
1    | /  macro_rules! int_impl {
2    | |      (
3    | |          Self = $SelfT:ty,
4    | |          ActualT = $ActualT:ident,
...    |
2749 | |              if intrinsics::is_val_statically_known(exp) {
...    |
3646 | |      }
3647 | |  }
     | |__- in this expansion of `int_impl!`
---
    --> library/core/src/num/int_macros.rs:2749:16
     |
1    | /  macro_rules! int_impl {
2    | |      (
3    | |          Self = $SelfT:ty,
4    | |          ActualT = $ActualT:ident,
...    |
2749 | |              if intrinsics::is_val_statically_known(exp) {
...    |
3646 | |      }
3647 | |  }
     | |__- in this expansion of `int_impl!`
---
    --> library/core/src/num/int_macros.rs:2749:16
     |
1    | /  macro_rules! int_impl {
2    | |      (
3    | |          Self = $SelfT:ty,
4    | |          ActualT = $ActualT:ident,
...    |
2749 | |              if intrinsics::is_val_statically_known(exp) {
...    |
3646 | |      }
3647 | |  }
     | |__- in this expansion of `int_impl!`
```

</details>

mzabaluev commented:
In https://github.com/rust-lang/rust/pull/122884/commits/010c332555dc0fd79d02c7b04a5cfac40f0af279, I tried to use `is_val_statically_known`. There is no hint (I checked with the nightly compiler this time), furthermore, enabling the feature in the library does not help.

oskgo commented:
Sounds like this would help: https://rustc-dev-guide.rust-lang.org/stability.html#rustc_allow_const_fn_unstable

mzabaluev commented:
> Sounds like this would help: https://rustc-dev-guide.rust-lang.org/stability.html#rustc_allow_const_fn_unstable

It's already enabled in the library.

Amanieu commented:
You need `#![feature(is_val_statically_known)]`.

Amanieu commented:
That's strange the feature seems to work fine here: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=761301ba758e4ba8d60381021abf908a

mzabaluev commented:
It is what it says on the tin: `pow` is annotated as const-stable, so it cannot call the const-unstable `is_val_statically_known`.
Your playground examples don't (can't) use stability attributes.

Amanieu commented:
Right, in that case maybe it's best to go back to the version with the unroll loop.

mzabaluev commented:
Oh, I get it: `rustc_allow_const_fn_unstable` is an item attribute that is enabled by the feature.

oskgo commented:
pinging @rust-lang/wg-const-eval due to new usage of `rustc_allow_const_fn_unstable`. It should be fine since this PR is purely an optimization and can always be reverted.

RalfJung commented:
`is_val_statically_known` is a very harmless intrinsic from a const-eval perspective, so seems fine for me.

Amanieu commented on library/core/src/num/int_macros.rs:
```diff
@@ -2174,26 +2172,57 @@ macro_rules! int_impl {
+                // This gives the optimizer a way to efficiently inline call sites
+                // for the most common use cases with constant exponents.
+                // Currently, LLVM is unable to unroll the loop below.
+                match exp {
```
Rather than this special casing could we instead just have the original loop (which LLVM knows how to unroll) for the `is_val_statically_known` case and your new loop for the non-constant case?

And do the same for all the other `pow` functions.

mzabaluev replied:
I have updated `pow` and `wrapped_pow` as suggested.
I'm not sure the extra complication is justified for the checked operations, but I guess the optimizer will have better opportunities with the original loop there as well. I will try to make a macro so that uniform code is used everywhere without repetition.

mzabaluev committed:
Revert to original loop for const pow exponents

Give LLVM the for original, optimizable loop in pow and wrapped_pow
functions in the case when the exponent is statically known.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10364099839/job/28688821775) [(plain)](https://github.com/rust-lang/rust/commit/10db28f1f8c8cd0c4426ec41f3f8e003d27d73d7/checks/28688821775/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.659 Building wheels for collected packages: reuse
#13 2.660   Building wheel for reuse (pyproject.toml): started
#13 2.911   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.912   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 2.912   Stored in directory: /tmp/pip-ephem-wheel-cache-pr2jz7zl/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.915 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.315 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.315 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.847 Collecting virtualenv
#13 3.847 Collecting virtualenv
#13 3.904   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 3.986      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 71.5 MB/s eta 0:00:00
#13 4.043 Collecting filelock<4,>=3.12.2
#13 4.047   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.065 Collecting distlib<1,>=0.3.7
#13 4.068   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.075      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 91.0 MB/s eta 0:00:00
#13 4.104 Collecting platformdirs<5,>=3.9.1
#13 4.108   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.189 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.379 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      202688 kB
DirectMap2M:     8185856 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/e9e27ab0cf3759e420eeaa6bab7c7d454d9c10cd/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-e9e27ab0cf3759e420eeaa6bab7c7d454d9c10cd-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
                     base = base.wrapping_mul(base);
                 }
-    
+
                 // since exp!=0, finally the exp must be 1.
                 // Deal with the final bit of the exponent separately, since
                 // squaring the base afterwards is not necessary.
Diff in /checkout/library/core/src/num/int_macros.rs:2189:
                     exp /= 2;
                     base = base.wrapping_mul(base);
-    
+
+
                 // since exp!=0, finally the exp must be 1.
                 // Deal with the final bit of the exponent separately, since
                 // squaring the base afterwards is not necessary.
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/core/src/borrow.rs" "/checkout/compiler/rustc_middle/src/lint.rs" "/checkout/library/core/src/pat.rs" "/checkout/library/core/src/char/decode.rs" "/checkout/library/core/src/char/convert.rs" "/checkout/library/core/src/char/mod.rs" "/checkout/library/core/src/char/methods.rs" "/checkout/library/core/src/fmt/num.rs" "/checkout/library/core/src/fmt/float.rs" "/checkout/library/core/src/fmt/mod.rs" "/checkout/library/core/src/fmt/builders.rs" "/checkout/library/core/src/fmt/rt.rs" "/checkout/library/core/src/fmt/nofloat.rs" "/checkout/library/core/src/intrinsics/simd.rs" "/checkout/library/core/src/intrinsics/mir.rs" "/checkout/library/core/src/cmp/bytewise.rs" "/checkout/library/core/src/convert/num.rs" "/checkout/library/core/src/convert/mod.rs" "/checkout/library/core/src/sync/mod.rs" "/checkout/library/core/src/sync/exclusive.rs" "/checkout/library/core/src/sync/atomic.rs" "/checkout/compiler/rustc_middle/src/ty/pattern.rs" "/checkout/compiler/rustc_middle/src/ty/typeck_results.rs" "/checkout/compiler/rustc_middle/src/ty/opaque_types.rs" "/checkout/compiler/rustc_middle/src/ty/context/tls.rs" "/checkout/compiler/rustc_middle/src/ty/walk.rs" "/checkout/compiler/rustc_middle/src/ty/context.rs" "/checkout/compiler/rustc_middle/src/ty/sty.rs" "/checkout/compiler/rustc_middle/src/ty/error.rs" "/checkout/compiler/rustc_middle/src/ty/consts.rs" "/checkout/compiler/rustc_middle/src/ty/predicate.rs" "/checkout/compiler/rustc_middle/src/ty/mod.rs" "/checkout/compiler/rustc_middle/src/ty/flags.rs" "/checkout/compiler/rustc_middle/src/ty/fast_reject.rs" "/checkout/compiler/rustc_middle/src/ty/layout.rs" "/checkout/compiler/rustc_middle/src/ty/list.rs" "/checkout/compiler/rustc_middle/src/ty/diagnostics.rs" "/checkout/compiler/rustc_middle/src/ty/generic_args.rs" "/checkout/compiler/rustc_middle/src/ty/adjustment.rs" "/checkout/compiler/rustc_middle/src/ty/closure.rs" "/checkout/compiler/rustc_middle/src/ty/fold.rs" "/checkout/compiler/rustc_middle/src/ty/normalize_erasing_regions.rs" "/checkout/compiler/rustc_middle/src/ty/vtable.rs" "/checkout/compiler/rustc_middle/src/ty/intrinsic.rs" "/checkout/compiler/rustc_middle/src/ty/consts/kind.rs" "/checkout/compiler/rustc_middle/src/ty/consts/valtree.rs" "/checkout/compiler/rustc_middle/src/ty/consts/int.rs" "/checkout/library/core/src/num/int_macros.rs" "/checkout/library/core/src/num/error.rs" "/checkout/library/core/src/num/mod.rs" "/checkout/library/core/src/num/f128.rs" "/checkout/compiler/rustc_middle/src/ty/print/mod.rs" "/checkout/compiler/rustc_middle/src/ty/print/pretty.rs" "/checkout/compiler/rustc_middle/src/ty/abstract_const.rs" "/checkout/compiler/rustc_middle/src/ty/impls_ty.rs" "/checkout/compiler/rustc_middle/src/ty/codec.rs" "/checkout/compiler/rustc_middle/src/ty/parameterized.rs" "/checkout/compiler/rustc_middle/src/ty/inhabitedness/mod.rs" "/checkout/compiler/rustc_middle/src/ty/inhabitedness/inhabited_predicate.rs" "/checkout/compiler/rustc_middle/src/ty/rvalue_scopes.rs" "/checkout/compiler/rustc_middle/src/ty/cast.rs" "/checkout/compiler/rustc_middle/src/ty/assoc.rs" "/checkout/compiler/rustc_middle/src/ty/visit.rs" "/checkout/library/core/src/time.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Tue Aug 13 05:30:21 UTC 2024
  network time: Tue, 13 Aug 2024 05:30:21 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:pushpin: Commit ac88b330b875e8058589b1804ac5d95fcd40905d has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu ac88b330b875e8058589b1804ac5d95fcd40905d -->
<!-- homu: {"type":"Approved","sha":"ac88b330b875e8058589b1804ac5d95fcd40905d","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
